<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Introduction to corncob, no phyloseq • corncob</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to corncob, no phyloseq">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">corncob</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/corncob-intro-no-phyloseq.html">Introduction to corncob, no phyloseq</a></li>
    <li><a class="dropdown-item" href="../articles/corncob-intro.html">Introduction to corncob</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/statdivlab/corncob/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Introduction to corncob, no phyloseq</h1>
                        <h4 data-toc-skip class="author">Bryan D
Martin</h4>
            
            <h4 data-toc-skip class="date">2025-03-17</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/statdivlab/corncob/blob/main/vignettes/corncob-intro-no-phyloseq.Rmd" class="external-link"><code>vignettes/corncob-intro-no-phyloseq.Rmd</code></a></small>
      <div class="d-none name"><code>corncob-intro-no-phyloseq.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="vignette-information">Vignette Information<a class="anchor" aria-label="anchor" href="#vignette-information"></a>
</h2>
<p>This is a version of the <code>corncob-intro.Rmd</code> vignette that
does not rely on the package <code>phyloseq</code>. To see how to use
<code>corncob</code> with <code>phyloseq</code> objects, or to see an
additional analysis with an IBD microbiome dataset, check out the
original <code>corncob-intro.Rmd</code> vignette.</p>
<p>We thank Dr. Thea Whitman for kindly providing us with the example
data set we use for this vignette. You can read more about this data in
Whitman, Thea, et al. “Dynamics of microbial community composition and
soil organic carbon mineralization in soil following addition of
pyrogenic and fresh organic matter.” <em>The ISME Journal</em> 10.12
(2016): 2918.</p>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
Effectively modeling microbial relative abundance poses a number of
statistical challenges, including:
<p>Here, we introduce <code>corncob</code>, an individual taxon
regression model that uses abundance tables and sample data.
<code>corncob</code> is able to model differential abundance and
differential variability, and addresses each of the challenges presented
above.</p>
<p>Install <code>corncob</code> using:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"statdivlab/corncob"</span><span class="op">)</span></span></code></pre></div>
<p>To begin, we load our example data set we load our example data set
as three different data frames</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/statdivlab/corncob" class="external-link">corncob</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">soil_phylo_sample</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">soil_phylo_otu</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">soil_phylo_taxa</span><span class="op">)</span></span></code></pre></div>
<p>We can see that we have 5 sample variables. Let’s look at the first
few observations.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">soil_phylo_sample</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      Plants DayAmdmt Amdmt ID Day</span></span>
<span><span class="co">## S009      1       01     1  D   0</span></span>
<span><span class="co">## S204      1       21     1  D   2</span></span>
<span><span class="co">## S112      0       11     1  B   1</span></span>
<span><span class="co">## S247      0       22     2  F   2</span></span>
<span><span class="co">## S026      0       00     0  A   0</span></span>
<span><span class="co">## S023      1       00     0  C   0</span></span></code></pre>
Our covariates are as follows:
<p>Additionally, we have a table that gives us counts of the phyla by
sample. Let’s take a look at the first 5 phyla and the first 5
samples.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">soil_phylo_otu</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##         S009 S204 S112 S247 S026</span></span>
<span><span class="co">## OTU.43   350   74  300   70   43</span></span>
<span><span class="co">## OTU.2   1796 4204 1752  695  945</span></span>
<span><span class="co">## OTU.187  280  709  426  100  139</span></span>
<span><span class="co">## OTU.150   33  151   18   13   28</span></span>
<span><span class="co">## OTU.91     0    0  184    0    0</span></span></code></pre>
<p>Finally, we have a taxonomy table with 7 taxonomic ranks.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">soil_phylo_taxa</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##         Kingdom    Phylum           Class                 Order             </span></span>
<span><span class="co">## OTU.43  "Bacteria" "Nitrospirae"    "Nitrospira"          "Nitrospirales"   </span></span>
<span><span class="co">## OTU.2   "Bacteria" "Proteobacteria" "Alphaproteobacteria" "Rhizobiales"     </span></span>
<span><span class="co">## OTU.187 "Bacteria" "Acidobacteria"  "Acidobacteriia"      "Acidobacteriales"</span></span>
<span><span class="co">##         Family              Genus            Species</span></span>
<span><span class="co">## OTU.43  "Nitrospiraceae"    "Nitrospira"     ""     </span></span>
<span><span class="co">## OTU.2   "Bradyrhizobiaceae" "Bradyrhizobium" ""     </span></span>
<span><span class="co">## OTU.187 "Koribacteraceae"   ""               ""</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="fitting-a-model">Fitting a Model<a class="anchor" aria-label="anchor" href="#fitting-a-model"></a>
</h2>
<p>First, let’s subset our samples to only include those with the
<code>DayAmdmt</code> covariate equal to 11 or 21 and then collapse the
samples to the phylum level. We have already done this and saved the
resulting sample data frame and otu data frame.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">soil_phylum_small_sample</span><span class="op">)</span></span>
<span><span class="va">sample_data</span> <span class="op">&lt;-</span> <span class="va">soil_phylum_small_sample</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">soil_phylum_small_otu</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">soil_phylum_small_otu</span></span></code></pre></div>
<p>Note that collapsing the samples is not necessary, and this model can
work at any taxonomic rank. However, we will later be fitting a model to
every taxa. We can see that by agglomerating taxa to the phylum level,
we have gone from from 7770 to 40 taxa. Thus we collapse in order to
increase the speed for the purposes of this tutorial.</p>
<p>Now we fit our model. We will demonstrate with Proteobacteria. We
will need to make a data frame that includes our sample data as well as
the Proteobacteria counts and sequencing depths for each sample.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pro_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">sample_data</span>, </span>
<span>                  W <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">data</span><span class="op">[</span><span class="st">"Proteobacteria"</span>, <span class="op">]</span><span class="op">)</span>,</span>
<span>                  M <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>For now, we will not include any covariates, so we use
<code>~ 1</code> as our model formula responses.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">corncob</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bbdml.html">bbdml</a></span><span class="op">(</span>formula <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">W</span>, <span class="va">M</span> <span class="op">-</span> <span class="va">W</span><span class="op">)</span> <span class="op">~</span> <span class="fl">1</span>,</span>
<span>             phi.formula <span class="op">=</span> <span class="op">~</span> <span class="fl">1</span>,</span>
<span>             data <span class="op">=</span> <span class="va">pro_data</span><span class="op">)</span></span></code></pre></div>
<div style="page-break-after: always;"></div>
</div>
<div class="section level2">
<h2 id="interpreting-a-model">Interpreting a Model<a class="anchor" aria-label="anchor" href="#interpreting-a-model"></a>
</h2>
<p>First, let’s plot the data with our model fit on the relative
abundance scale. To do this, we simply type:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">corncob</span>, B <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<p><img src="corncob-intro-no-phyloseq_files/figure-html/unnamed-chunk-10-1.png" width="768"></p>
<p>You can access the documentation for this plotting function by typing
<code><a href="../reference/plot.bbdml.html">?plot.bbdml</a></code> into the console.</p>
<p>The points represent the relative abundances. The bars represent the
95% prediction intervals for the observed relative abundance by sample.
The parameter <code>B</code> determines the number of bootstrap
simulations used to approximate the prediction intervals. For purposes
of this tutorial, we use a small value <code>B = 50</code> for
computational purposes, but recommend a higher setting for more accurate
intervals, such as the default <code>B = 1000</code>.</p>
<p>Now let’s look at the same plot, but on the counts scale with 95%
prediction intervals (since counts is not a parameter). To do this, we
add the option <code>total = TRUE</code> to our plotting code.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">corncob</span>, total <span class="op">=</span> <span class="cn">TRUE</span>, B <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<p><img src="corncob-intro-no-phyloseq_files/figure-html/unnamed-chunk-11-1.png" width="768"></p>
<p>Finally, let’s color the plot by the <code>DayAmdmt</code> covariate.
To do this, we add the option <code>color = "DayAmdmt"</code> to our
plotting code.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">corncob</span>, total <span class="op">=</span> <span class="cn">TRUE</span>, color <span class="op">=</span> <span class="st">"DayAmdmt"</span>, B <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<p><img src="corncob-intro-no-phyloseq_files/figure-html/unnamed-chunk-12-1.png" width="768"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">corncob</span>, color <span class="op">=</span> <span class="st">"DayAmdmt"</span>, B <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<p><img src="corncob-intro-no-phyloseq_files/figure-html/unnamed-chunk-13-1.png" width="768"></p>
<p>Notice that this plot also reorders our samples so that groups appear
together so that they are easier to compare.</p>
<p>We can observe on this plot that it might be of interest to
distinguish between the two groups with covariates. The average
empirical relative abundance for the samples with
<code>DayAmdmt = 21</code> tends to be lower and less variable than the
samples with <code>DayAmdmt = 11</code>.</p>
</div>
<div class="section level2">
<h2 id="adding-covariates">Adding covariates<a class="anchor" aria-label="anchor" href="#adding-covariates"></a>
</h2>
<p>Let’s try modeling the expected relative abundance and the
variability of the counts with <code>DayAmdmt</code> as a covariate. We
do this by modifying <code>formula</code> and <code>phi.formula</code>
as:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">corncob_da</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bbdml.html">bbdml</a></span><span class="op">(</span>formula <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">W</span>, <span class="va">M</span> <span class="op">-</span> <span class="va">W</span><span class="op">)</span> <span class="op">~</span> <span class="va">DayAmdmt</span>,</span>
<span>             phi.formula <span class="op">=</span> <span class="op">~</span> <span class="va">DayAmdmt</span>,</span>
<span>             data <span class="op">=</span> <span class="va">pro_data</span><span class="op">)</span></span></code></pre></div>
<p>Let’s also plot this data on both the total count and relative
abundance scales.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">corncob_da</span>, color <span class="op">=</span> <span class="st">"DayAmdmt"</span>, total <span class="op">=</span> <span class="cn">TRUE</span>, B <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<p><img src="corncob-intro-no-phyloseq_files/figure-html/unnamed-chunk-15-1.png" width="768"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">corncob_da</span>, color <span class="op">=</span> <span class="st">"DayAmdmt"</span>, B <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<p><img src="corncob-intro-no-phyloseq_files/figure-html/unnamed-chunk-16-1.png" width="768"></p>
<p>Visually, the model with covariates seems to provide a much better
fit to the data, but how can we compare the two models
statistically?</p>
</div>
<div class="section level2">
<h2 id="model-selection">Model Selection<a class="anchor" aria-label="anchor" href="#model-selection"></a>
</h2>
<p>Let’s use a likelihood ratio test to select our final model for this
taxon. We want to test the null hypothesis that the likelihood of the
model with covariates is equal to the likelihood of the model without
covariates. To do this test, we use:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/lrtest.html">lrtest</a></span><span class="op">(</span>mod_null <span class="op">=</span> <span class="va">corncob</span>, mod <span class="op">=</span> <span class="va">corncob_da</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 4.550571e-05</span></span></code></pre>
<p>We obtain a p-value much smaller than a cut-off of 0.05. Therefore we
conclude that there is a statistically significant difference in the
likelihood of the two models. Thus, we probably want to use the model
with covariates for this taxon.</p>
</div>
<div class="section level2">
<h2 id="parameter-interpretation">Parameter Interpretation<a class="anchor" aria-label="anchor" href="#parameter-interpretation"></a>
</h2>
<p>Now that we have chosen our model, let’s interpret our model output.
To see a summary of the model, type:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">corncob_da</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Call:</span></span>
<span><span class="co">## bbdml(formula = cbind(W, M - W) ~ DayAmdmt, phi.formula = ~DayAmdmt, </span></span>
<span><span class="co">##     data = pro_data)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Coefficients associated with abundance:</span></span>
<span><span class="co">##             Estimate Std. Error t value Pr(&gt;|t|)    </span></span>
<span><span class="co">## (Intercept) -0.44595    0.03604 -12.375 7.18e-13 ***</span></span>
<span><span class="co">## DayAmdmt21  -0.16791    0.04067  -4.129 0.000297 ***</span></span>
<span><span class="co">## ---</span></span>
<span><span class="co">## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Coefficients associated with dispersion:</span></span>
<span><span class="co">##             Estimate Std. Error t value Pr(&gt;|t|)    </span></span>
<span><span class="co">## (Intercept)  -5.3077     0.3537 -15.008 6.44e-15 ***</span></span>
<span><span class="co">## DayAmdmt21   -1.3518     0.5029  -2.688    0.012 *  </span></span>
<span><span class="co">## ---</span></span>
<span><span class="co">## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Log-likelihood: -286.53</span></span></code></pre>
<p>This output will look familiar if you have done regression analysis
in R in the past. Covariates associated with the expected relative
abundance are presented separately from covariates associated with the
variance of the counts are preceded by.</p>
<p>From this model summary, we can see that the <code>DayAmdmt21</code>
abundance coefficient is negative and statistically significant. This
suggests that this taxon is differentially-abundant across
<code>DayAmdmt</code>, and that samples with <code>DayAmdmt = 21</code>
are expected to have a lower relative abundance. This matches what we
saw from the observed abundances.</p>
<p>We can also see that the <code>DayAmdmt21</code> dispersion
coefficient is negative and statistically significant. This suggests
that this taxon is differentially-variable across <code>DayAmdmt</code>,
and that samples with <code>DayAmdmt = 21</code> are expected to have a
lower variability. This matches what we saw from the observed
abundances.</p>
</div>
<div class="section level2">
<h2 id="analysis-for-multiple-taxa">Analysis for Multiple Taxa<a class="anchor" aria-label="anchor" href="#analysis-for-multiple-taxa"></a>
</h2>
<p>What if we want to test all the taxa in our data to see if they are
differentially-abundant or differentially-variable? We use the
<code>differentialTest</code> function. It will perform the above tests
on all taxa, and it will control the false discovery rate to account for
multiple comparisons.</p>
<p>Next, we use the <code>differentialTest</code> command. We specify
the covariates of our model using <code>formula</code> and
<code>phi.formula</code> as before, except we no longer include the
response term because we are testing multiple taxa. We also specify
which covariates we want to test for by removing them in the
<code>formula_null</code> and <code>phi.formula_null</code>
arguments.</p>
<p>The difference between the formulas and the null version of the
formulas are the variables that we test. We will go through several
examples, starting with a test for differential abundance across the
<code>DayAmdmt</code> coefficient.</p>
<p>We set <code>fdr_cutoff</code> to be our controlled false discovery
rate.</p>
<p>Here, we will pass in our entire count table for all phyla as the
<code>data</code> argument (setting <code>taxa_are_rows</code> to
<code>TRUE</code>), and the sample data frame as the
<code>sample_data</code> argument.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">da_analysis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/differentialTest.html">differentialTest</a></span><span class="op">(</span>formula <span class="op">=</span> <span class="op">~</span> <span class="va">DayAmdmt</span>,</span>
<span>                                 phi.formula <span class="op">=</span> <span class="op">~</span> <span class="va">DayAmdmt</span>,</span>
<span>                                 formula_null <span class="op">=</span> <span class="op">~</span> <span class="fl">1</span>,</span>
<span>                                 phi.formula_null <span class="op">=</span> <span class="op">~</span> <span class="va">DayAmdmt</span>,</span>
<span>                                 test <span class="op">=</span> <span class="st">"Wald"</span>, boot <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                 data <span class="op">=</span> <span class="va">data</span>,</span>
<span>                                 sample_data <span class="op">=</span> <span class="va">sample_data</span>,</span>
<span>                                 taxa_are_rows <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                 fdr_cutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<p>We can see the output of the function by calling it:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">da_analysis</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class differentialTest </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $p: p-values </span></span>
<span><span class="co">## $p_fdr: FDR-adjusted p-values </span></span>
<span><span class="co">## $significant_taxa: taxa names of the statistically significant taxa </span></span>
<span><span class="co">## $significant_models: model summaries of the statistically significant taxa </span></span>
<span><span class="co">## $all_models: all model summaries </span></span>
<span><span class="co">## $restrictions_DA: covariates tested for differential abundance </span></span>
<span><span class="co">## $restrictions_DV: covariates tested for differential variability </span></span>
<span><span class="co">## $discriminant_taxa_DA: taxa for which at least one covariate associated with the abundance was perfectly discriminant </span></span>
<span><span class="co">## $discriminant_taxa_DV: taxa for which at least one covariate associated with the dispersion was perfectly discriminant </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## plot( ) to see a plot of tested coefficients from significant taxa</span></span></code></pre>
<p>We can see a list of differentially-abundant taxa using:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">da_analysis</span><span class="op">$</span><span class="va">significant_taxa</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "Proteobacteria"   "Gemmatimonadetes" "Bacteroidetes"    "Cyanobacteria"   </span></span>
<span><span class="co">##  [5] "Firmicutes"       "Planctomycetes"   "Armatimonadetes"  "Spirochaetes"    </span></span>
<span><span class="co">##  [9] "Elusimicrobia"    "BRC1"             "OP3"              "FBP"             </span></span>
<span><span class="co">## [13] "Chlorobi"         "TM6"</span></span></code></pre>
<p>In this case, we identified 14 taxa that are differentially-abundant
across <code>DayAmdmt</code> (out of the 39 taxa tested).</p>
<p>We can see a list of differentially-variable taxa using:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">dv_analysis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/differentialTest.html">differentialTest</a></span><span class="op">(</span>formula <span class="op">=</span> <span class="op">~</span> <span class="va">DayAmdmt</span>,</span>
<span>                                 phi.formula <span class="op">=</span> <span class="op">~</span> <span class="va">DayAmdmt</span>,</span>
<span>                                 formula_null <span class="op">=</span> <span class="op">~</span> <span class="va">DayAmdmt</span>,</span>
<span>                                 phi.formula_null <span class="op">=</span> <span class="op">~</span> <span class="fl">1</span>,</span>
<span>                                 test <span class="op">=</span> <span class="st">"LRT"</span>, boot <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                 data <span class="op">=</span> <span class="va">data</span>,</span>
<span>                                 sample_data <span class="op">=</span> <span class="va">sample_data</span>,</span>
<span>                                 taxa_are_rows <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                 fdr_cutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="va">dv_analysis</span><span class="op">$</span><span class="va">significant_taxa</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Acidobacteria" "Cyanobacteria" "Spirochaetes"  "Elusimicrobia"</span></span>
<span><span class="co">## [5] "FBP"</span></span></code></pre>
<p>In this case, we identified 5 taxa that are differentially-variable
across <code>DayAmdmt</code> (out of the 40 taxa tested).</p>
<p>We can examine a subset of the p-values of our tests using:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">da_analysis</span><span class="op">$</span><span class="va">p</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##    Acidobacteria   Proteobacteria Gemmatimonadetes   Actinobacteria </span></span>
<span><span class="co">##     6.509417e-01     3.642734e-05     3.270448e-13     3.703095e-01 </span></span>
<span><span class="co">##         [Thermi] </span></span>
<span><span class="co">##     1.031419e-01</span></span></code></pre>
<p>We can examine a subset of the p-values after controlling for the
false discovery rate using:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">da_analysis</span><span class="op">$</span><span class="va">p_fdr</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##    Acidobacteria   Proteobacteria Gemmatimonadetes   Actinobacteria </span></span>
<span><span class="co">##     7.811301e-01     1.457094e-04     3.924537e-12     5.172748e-01 </span></span>
<span><span class="co">##         [Thermi] </span></span>
<span><span class="co">##     2.062838e-01</span></span></code></pre>
<p>where the values are now adjusted to control the false discovery rate
at 0.05.</p>
<p>Now we can use the built-in plotting function in corncob to view the
the model coefficients used in testing differential abundance across
<code>DayAmdmt</code> via the <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> function. You can
access the documentation for this plotting function by typing
<code><a href="../reference/plot.differentialTest.html">?plot.differentialTest</a></code> into the console.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">da_analysis</span><span class="op">)</span></span></code></pre></div>
<p><img src="corncob-intro-no-phyloseq_files/figure-html/unnamed-chunk-25-1.png" width="768"></p>
<p>Here, we can see that for <code>Bacteria_Armatimonadetes</code>, the
effect of <code>DayAmdmt21</code> is positive compared to the baseline
(in this case, <code>DayAmdmt11</code>).</p>
<p>If you wish to instead make your own custom plots with these same
coefficients, you can easily extract the data used to generate the plots
above from the <code>plot</code> function by setting
<code>data_only = TRUE</code>:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">da_analysis</span>, data_only <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># we can easily remove special characters used in our formatting steps</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="va">df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>variable <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"\nDifferential Abundance"</span>, <span class="st">""</span>,</span>
<span>                                <span class="va">variable</span>, fixed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            x        xmin        xmax             taxa   variable</span></span>
<span><span class="co">## 1 -0.1679129 -0.24761731 -0.08820844   Proteobacteria DayAmdmt21</span></span>
<span><span class="co">## 2  0.3119623  0.22800588  0.39591870 Gemmatimonadetes DayAmdmt21</span></span>
<span><span class="co">## 3  0.2306765  0.09148981  0.36986312    Bacteroidetes DayAmdmt21</span></span>
<span><span class="co">## 4  1.7090454  1.33732670  2.08076404    Cyanobacteria DayAmdmt21</span></span>
<span><span class="co">## 5 -0.3087530 -0.49665466 -0.12085125       Firmicutes DayAmdmt21</span></span>
<span><span class="co">## 6  0.2013158  0.11895313  0.28367838   Planctomycetes DayAmdmt21</span></span></code></pre>
<p>Finally, we can see a list of any taxa for which we were not able to
fit a model using:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">da_analysis</span><span class="op">$</span><span class="va">p</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="va">names</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "GN04"   "GN02"   "MVP-21"</span></span></code></pre>
<p>In this case, we weren’t able to fit <code>GN04</code> automatically.
It’s worthwhile to investigate the OTU individually if this is the
case.</p>
<p>It may be that the model is overparameterized because there aren’t
enough observations, or it may just be that the initializations were
invalid for that taxa and it needs to be re-evaluated with new
initializations.</p>
<p>Let’s first try examining the data.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span><span class="op">[</span><span class="st">"GN04"</span>, <span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##      S204 S112 S134 S207 S202 S139 S122 S212 S117 S104 S214 S109 S217 S229 S132</span></span>
<span><span class="co">## GN04    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">##      S209 S227 S107 S237 S224 S127 S137 S114 S124 S119 S219 S232 S129 S102 S234</span></span>
<span><span class="co">## GN04    0    0    0    0    0    0    0    0    0    0    0    0    0    1    0</span></span>
<span><span class="co">##      S222 S239</span></span>
<span><span class="co">## GN04    0    0</span></span></code></pre>
<p>We see that the observed counts of OTU is zero in all samples except
for <code>S102</code>, where we observed a single count. We should be
skeptical of <strong>any</strong> statistical model fit on a single
observed count!</p>
<p><code>corncob</code> is stable, but if you notice any issues, please
<a href="https://github.com/statdivlab/corncob/issues" class="external-link">log them on
Github</a> to help us help you!</p>
</div>
<div class="section level2">
<h2 id="examples-of-answering-scientific-questions">Examples of Answering Scientific Questions<a class="anchor" aria-label="anchor" href="#examples-of-answering-scientific-questions"></a>
</h2>
<p>We will now walk through several scientific questions of interest and
show how they can be answered using hypothesis testing with
<code>corncob</code>. Note that <code>Day</code> and <code>Amdmt</code>
are both factor covariates with levels 0, 1, and 2.</p>
<p>Note that some of these are rather strange tests, and shown for
demonstration of the flexibility of the model only. Normally, when
testing for differential variability across a covariate, we recommend
always controlling for the effect of that covariate on the abundance. We
first demonstrate examples with the small version of the soil dataset
(note, this is a slightly different subset of the data from that used in
<code>corncob-intro.Rmd</code>).</p>
<p>Testing for differential abundance across <code>Day</code>, without
controlling for anything else:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ex1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/differentialTest.html">differentialTest</a></span><span class="op">(</span>formula <span class="op">=</span> <span class="op">~</span> <span class="va">Day</span>,</span>
<span>                        phi.formula <span class="op">=</span> <span class="op">~</span> <span class="fl">1</span>,</span>
<span>                        formula_null <span class="op">=</span> <span class="op">~</span> <span class="fl">1</span>,</span>
<span>                        phi.formula_null <span class="op">=</span> <span class="op">~</span> <span class="fl">1</span>,</span>
<span>                        data <span class="op">=</span> <span class="va">data</span>,</span>
<span>                        taxa_are_rows <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                        sample_data <span class="op">=</span> <span class="va">sample_data</span>, </span>
<span>                        test <span class="op">=</span> <span class="st">"Wald"</span>, boot <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                        fdr_cutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">ex1</span><span class="op">)</span></span></code></pre></div>
<p><img src="corncob-intro-no-phyloseq_files/figure-html/unnamed-chunk-29-1.png" width="768"></p>
<p>Testing for differential abundance across <code>Day</code>,
controlling for the effect of <code>Day</code> on dispersion:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ex2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/differentialTest.html">differentialTest</a></span><span class="op">(</span>formula <span class="op">=</span> <span class="op">~</span> <span class="va">Day</span>,</span>
<span>                        phi.formula <span class="op">=</span> <span class="op">~</span> <span class="va">Day</span>,</span>
<span>                        formula_null <span class="op">=</span> <span class="op">~</span> <span class="fl">1</span>,</span>
<span>                        phi.formula_null <span class="op">=</span> <span class="op">~</span> <span class="va">Day</span>,</span>
<span>                        data <span class="op">=</span> <span class="va">data</span>,</span>
<span>                        taxa_are_rows <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                        sample_data <span class="op">=</span> <span class="va">sample_data</span>, </span>
<span>                        test <span class="op">=</span> <span class="st">"Wald"</span>, boot <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                        fdr_cutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">ex2</span><span class="op">)</span></span></code></pre></div>
<p><img src="corncob-intro-no-phyloseq_files/figure-html/unnamed-chunk-30-1.png" width="768"></p>
<p>Jointly testing for differential abundance and differential
variability across <code>Day</code>:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ex3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/differentialTest.html">differentialTest</a></span><span class="op">(</span>formula <span class="op">=</span> <span class="op">~</span> <span class="va">Day</span>,</span>
<span>                        phi.formula <span class="op">=</span> <span class="op">~</span> <span class="va">Day</span>,</span>
<span>                        formula_null <span class="op">=</span> <span class="op">~</span> <span class="fl">1</span>,</span>
<span>                        phi.formula_null <span class="op">=</span> <span class="op">~</span> <span class="fl">1</span>,</span>
<span>                        data <span class="op">=</span> <span class="va">data</span>,</span>
<span>                        taxa_are_rows <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                        sample_data <span class="op">=</span> <span class="va">sample_data</span>, </span>
<span>                        test <span class="op">=</span> <span class="st">"Wald"</span>, boot <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                        fdr_cutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">ex3</span><span class="op">)</span></span></code></pre></div>
<p><img src="corncob-intro-no-phyloseq_files/figure-html/unnamed-chunk-31-1.png" width="768"></p>
<p>More examples of answering scientific questions with a larger subset
of this soil data set and with an IBD data set can be seen in
<code>corncob-intro.Rmd</code>.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bryan D Martin, Daniela Witten, Amy D Willis.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
